Riot/AppDelegate.m

```diff
@@ -1120,6 +1156,70 @@ - (void)application:(UIApplication *)application didRegisterUserNotificationSett
+- (void)userNotificationCenter:(UNUserNotificationCenter *)center didReceiveNotificationResponse:(UNNotificationResponse *)response withCompletionHandler:(void (^)(void))completionHandler
+{
+    UNNotification *notification = response.notification;
+    UNNotificationContent *content = notification.request.content;
+    if ([[response actionIdentifier] isEqualToString:@"inline-reply"]) {
+        UNTextInputNotificationResponse *textResponse = (UNTextInputNotificationResponse *) response;
+        NSString* roomId = content.userInfo[@"room_id"];
+
+        if (roomId.length) {
+            NSArray* mxAccounts = [MXKAccountManager sharedManager].activeAccounts;
+
+            MXKRoomDataSourceManager* manager;
+            for (MXKAccount* account in mxAccounts)
+            {
+                MXRoom* room = [account.mxSession roomWithRoomId:roomId];
+                if (room)
+                {
+                    manager = [MXKRoomDataSourceManager sharedManagerForMatrixSession:account.mxSession];
+                    if (manager)
+                    {
+                        break;
+                    }
+                }
+            }
+            if (manager == nil)
+            {
+                NSLog(@"[AppDelegate][Push] didReceiveNotificationResponse: room with id %@ not found", roomId);
+            }
+            else
+            {
+                [manager roomDataSourceForRoom:roomId create:YES onComplete:^(MXKRoomDataSource *roomDataSource) {
+                    NSString* responseText = [textResponse userText];
+                    if (responseText != nil && responseText.length != 0)
+                    {
+                        NSLog(@"[AppDelegate][Push] didReceiveNotificationResponse: sending message to room: %@", roomId);
+                        [roomDataSource sendTextMessage:responseText success:^(NSString* eventId) {} failure:^(NSError* error) {
+                            UNMutableNotificationContent *failureNotificationContent = [[UNMutableNotificationContent alloc] init];
+                            failureNotificationContent.userInfo = content.userInfo;
+                            failureNotificationContent.body = NSLocalizedStringFromTable(@"room_event_failed_to_send", @"Vector", nil);
+
+                            UNNotificationRequest *failureNotificationRequest = [UNNotificationRequest
+                                    requestWithIdentifier:@"failureNotification"
+                                                  content:failureNotificationContent
+                                                  trigger:nil];
+
+                            [center addNotificationRequest:failureNotificationRequest withCompletionHandler:nil];
+                            NSLog(@"[AppDelegate][Push] didReceiveNotificationResponse: error sending text message: %@", error);
+                        }];
+                    }
+
+                    completionHandler();
+                }];
+            }
```

Comment:
You could add a `return;` at the end of this else like in `application:handleActionWithIdentifier: forLocalNotification:withResponseInfo:completionHandler:` to be sure `completionHandler();` is called only once.

Reply:
https://github.com/vector-im/riot-ios/commit/2f1b6153a8190ea4024db8cdb243b8d4338df5ea